README.md

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/ValeriVoev/danstat.svg?branch=master)](https://travis-ci.org/ValeriVoev/danstat)
[![CRAN
status](https://www.r-pkg.org/badges/version/danstat)](https://CRAN.R-project.org/package=danstat)
![Since when](https://www.r-pkg.org/badges/ago/danstat)
<!-- ![Downloads](https://cranlogs.r-pkg.org/badges/danstat) -->
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/danstat)
<!-- badges: end -->

## Installation

From CRAN:

``` r
install.packages("danstat")
```

From Github:

``` r
# install.packages("devtools")
devtools::install_github("ValeriVoev/danstat")
```

## Overview

The `danstat` package provides an R interface to Danmarks Statistik
Statistikbank API to enable an easier access to the wealth of data in
the data bank for research and the general community. The documentation
of the API can be found here: [Databank
API](https://www.dst.dk/en/Statistik/brug-statistikken/muligheder-i-statistikbanken/api).

## Usage

The API has 4 endpoints which are mimicked by four main functions of the
package:

1.  `get_subjects()` (SUBJECTS endpoint) retrieves information about
    subjects around which the data tables in the data bank are
    organized. The subjects are arranged hierarchically highest level
    like “Labour and income”, “Transport”, etc.
    `get_subjects()` retrieves the highest level of the hierarchy. See
    the function documentation for more details.
2.  `get_tables()` (TABLES endpoint) retrieves a list of tables
    associated with a given subject code. For example
    `get_tables(subjects = "2")` retrieves all tables related to the
    subject “Labour and income” with table id, description,
    variables in the table, etc.
3.  `get_table_metadata()` (TABLEINFO endpoint) returns information
    about a particular table - description, time of last update, whether
    or not it is actively updated, and most importantly (for practical
    purposes) the variable names and id’s which are needed whenever you
    request actual data from the table. Set `variables_only = TRUE` if
    you only want to get information on the table variables.
4.  `get_data()` (DATA endpoint) - returns data from a selected table.
    It is required to include a `variables` argument as a list. Each
    element of the list should itself be a named list (with elements
    `code` and `values`) where `code` is the variable id for which data
    is requested, and `values` is a vector of values for this variable.
    If all values are requested, specify `values = NA`. For example:

<!-- end list -->

``` r
library(danstat)
user_input = list(list(code = "ieland", values = c(5100, 5128)),
                  list(code = "køn", values = c(1,2)),
                  list(code = "tid", values = NA))

get_data(table_id = "folk1c", variables = user_input)
#> # A tibble: 192 x 4
#>   IELAND  KØN   TID    INDHOLD
#>   <chr>   <chr> <chr>    <dbl>
#> 1 Denmark Men   2008Q1 2465810
#> 2 Denmark Men   2008Q2 2466036
#> 3 Denmark Men   2008Q3 2467712
#> 4 Denmark Men   2008Q4 2469977
#> 5 Denmark Men   2009Q1 2470457
#> # … with 187 more rows
```

Note that while default language is set to English and variable values
are indeed returned in English, e.g. “Men”, column names are returned in
Danish, e.g. “KØN”, “INDHOLD”, etc. Unfortunately, the API doesn’t
currently provide an option to return column names (variable names) in
English. However, you can get the English translation using
`get_table_metadata`. For example, for the above table

``` r
library(dplyr)
get_table_metadata(table_id = "folk1c", variables_only = TRUE) %>% 
    select(id, text)
#>         id              text
#> 1   OMRÅDE            region
#> 2      KØN               sex
#> 3    ALDER               age
#> 4 HERKOMST          ancestry
#> 5   IELAND country of origin
#> 6      Tid              time
```

we can see that “Område” translates to “region”, “Køn” to “sex”, “Alder”
to “age”, etc. “Indhold” is always the “value” column whenever data is
returned with the `get_data` function.

## See also

There are (as far as I know) two other packages with similar
functionality:

  - <https://github.com/rOpenGov/dkstat>
  - <https://github.com/mikkelkrogsholm/statsDK>

In the packages above, the API is called with a `GET` request, while
`POST` is the prefrerred option of the API developers and is also what
is used in this package. Also, I think that using `POST` requests makes
the package code more readable compared to the long url-encoded queries
needed for `GET` requests. Also, as of this moment, the rOpenGov package
seems to not have been maintained for the past 3 years. In any case,
users can consider the above 2 packages as alternatives to this one.


NEWS.md
## Version 0.1.0

* First version available on Github
* Main package functionality released

## Version 0.2.0

* Package updated after changes to the Statistics Denmark API and subject hierarchy
* Available on CRAN again


tests/testthat.R
library(testthat)
library(danstat)

test_check("danstat")


tests/testthat/test-get_data.R
user_input_bad_code = list(
	list(code = "ieland", values = c(5100, 5128)),
	list(code = "køn", values = c(1,2)),
	list(code = "tid", values = NA),
	list(code = "not_in_table", values = c(2,3)))

user_input_bad_values = list(
	list(code = "køn", values = c(1,2)),
	list(code = "ieland", values = c(17, 5100, 5128)), # 17 is not a valid value
	list(code = "tid", values = NA))

user_input = list(list(code = "ieland", values = c(5100, 5128)),
									list(code = "køn", values = c(1,2)),
									list(code = "tid", values = NA))

test_that("error behavior with wrong and correct user input", {
  expect_error(check_variables_code("folk1c", user_input = user_input_bad_code))
	expect_true(check_variables_code("folk1c", user_input = user_input_bad_values)) # doesn't detect bad values, only bad codes
	expect_true(check_variables_code("folk1c", user_input = user_input))
	expect_true(check_variables_values(get_valid_variable_values("folk1c", "alder"), c("0-4", "5-9", "20-24")))
	expect_false(check_variables_values(get_valid_variable_values("folk1c", "alder"), c("0-4", "5-9", "20-24", "100+"))) # 100+ is not a valid value
})

test_that("error with wrong variable code", {
	expect_error(get_data(table_id = "folk1c", variables = user_input_bad_code))
})

test_that("warning with wrong variable values", {
	expect_warning(get_data(table_id = "folk1c", variables = user_input_bad_values))
})

test_that("correct column names and output class", {
	expect_equal(tolower(colnames(get_data(table_id = "folk1c", variables = user_input))), c("ieland", "køn", "tid", "indhold"))
	expect_is(get_data(table_id = "folk1c", variables = user_input), "data.frame")
})


tests/testthat/test-get_tables.R
test_that("subject 'as is' I() works", {
  expect_identical(get_tables(subjects = "3465"), get_tables(subjects = c("3465")))
})

test_that("numeric pastdays", {
	expect_error(get_tables(pastdays = "a"))
})


R/get_tables.R
#' Get a list of stables in the data bank
#'
#' @param subjects Provide specific subject id's to get subtopics. E.g. \code{subjects = c("02", "2419")}. Can be retrieved with \code{get_subjects()}
#' @param pastdays Return only tables which have been updated within this number of days
#' @param include_inactive Whether to return tables that are no longer updated
#' @param language Language for the return object. Default = \code{"en"}
#'
#' @importFrom httr POST content
#' @importFrom jsonlite fromJSON
#'
#' @return A data frame
#' @export
#' @examples
#' # Get all tables
#' all_tables <- get_tables()
#'
#' # Or get tables for specific subjects
#' some_tables <- get_tables(subjects = c("2", "3413"))
#'
#' # Get all tables updated within the past 3 days
#' tables_past3days <- get_tables(pastdays = 3)


# TODO: Consider implementing a search_term or regex to do a lookup in the "text" column which is a kind of description
get_tables <- function(subjects = NULL, pastdays = NA_integer_, include_inactive = FALSE, language = c("en", "da")){

	# evaluate language choices
	language <- match.arg(language)

	stopifnot(is.numeric(pastdays))

	# Treat objects "as is"
	if (!is.null(subjects)) subjects <- I(subjects)

	call_body <- list(lang = language,
										subjects = subjects,
										pastdays = pastdays,
										includeinactive = include_inactive)

	result <- httr::POST(TABLES_ENDPOINT, body = call_body, encode = "json")

	check_http_type(result)

	return(jsonlite::fromJSON(httr::content(result)))
}


R/get_subjects.R
#' Get a list of subjects covered in the data bank
#'
#' @param subjects Provide specific subject id's to get subtopics. E.g. \code{subjects = c("02", "2419")}
#' @param recursive Whether subtopics/tables will be retrieved all the way down the hierarchy. Otherwise, only the closest level under the provided subjects will be retrieved. Default = \code{FALSE}
#' @param include_tables Whether the result should contain tables. Otherwise, only subjects are returned. Default = \code{FALSE}
#' @param language Language for the return object. Default = \code{"en"}
#'
#'
#' @importFrom httr POST content
#' @importFrom jsonlite fromJSON
#'
#' @return A data frame
#' @export
#' @examples
#' # Get all subjects
#' all_subjects <- get_subjects()
#'
#' # Or get (sub)subjects for specific subjects
#' some_subjects <- get_subjects(subjects = c("2", "3"))
#'
#' # Get all subject hierarchy for a given subject
#' subject_with_hierarchy <- get_subjects(subjects = "2", recursive = TRUE)


get_subjects <- function(subjects = NULL, recursive = FALSE, include_tables = FALSE, language = c("en", "da")){

	# evaluate language choices
	language <- match.arg(language)
	# Treat objects "as is"
	if (!is.null(subjects)) subjects <- I(subjects)

	call_body <- list(lang = language,
										recursive = recursive,
										includeTables = include_tables,
										subjects = subjects)

	result <- httr::POST(SUBJECTS_ENDPOINT, body = call_body, encode = "json")

	check_http_type(result)

	return(jsonlite::fromJSON(httr::content(result)))
}



R/get_table_metadata.R
#' Title
#'
#' @param table_id Table identifier, e.g. "folk1a"
#' @param variables_only If \code{TRUE} returns only information about the variables in the table
#' @param language Language for the return object. Default = \code{"en"}
#'
#' @return A list with information about the table, like documentation url, variable description, etc. If \code{variables_only = TRUE}, returns a data frame with variable information.
#' @export
#' @examples
#' # Get table metadata for a given table
#' table_meta <- get_table_metadata(table_id = "folk1c") # a list
#'
#' # Get only information about the variables in the table
#' table_meta_vars <- get_table_metadata(table_id = "folk1c", variables_only = TRUE) # a data frame


get_table_metadata <- function(table_id, variables_only = FALSE, language = c("en", "da")){

	# evaluate language choices
	language <- match.arg(language)

	call_body <- list(lang = language,
										table = table_id)

	result <- httr::POST(METADATA_ENDPOINT, body = call_body, encode = "json")

	check_http_type(result)

	full_result <- jsonlite::fromJSON(httr::content(result))

	if (variables_only) return(full_result$variables)

	return(full_result)

}

get_valid_variable_values <- function(table_id, variable_id){

	vars <- get_table_metadata(table_id = table_id, variables_only = TRUE)
	return(vars[["values"]][[which(tolower(vars$id) == tolower(variable_id))]]$id)

}


R/get_data.R
#' Get data for a particular table and variable selection
#'
#' @param table_id Table identifier, e.g. "folk1a"
#' @param variables A list with variable code-values pairs. Each code-values pair should be a named list with names "code" and "values".
#' If all values for a variable are desired, define \code{values = NA} for that variable code.
#' @param language Language for the return object. Default = \code{"en"}
#'
#' @importFrom httr POST content
#' @importFrom jsonlite fromJSON
#' @importFrom readr read_csv2 locale
#'
#' @return A data frame
#' @export
#' @examples
#' # Get data from table "folk1c" for selected values of variables "ieland" and "køn"
#' # and all time periods available.
#'
#' # The "ieland" variable is filtered for Denmark (id = 5000) and Bulgaria (id = 5128)
#' # and the "køn" variable id filtered for Men (id = 1) and Women (id = 2).
#' # The "tid" variable is unfiltered, i.e. selects all available time periods
#' # See get_table_metadata(table_id = "folk1c", variables_only = TRUE) for variable codes and values.
#'
#' variables <- list(list(code = "ieland", values = c(5100, 5128)),
#'                   list(code = "køn", values = c(1,2)),
#'                   list(code = "tid", values = NA))
#'
#' data <- get_data("folk1c", variables)


get_data <- function(table_id, variables, language = c("en", "da")){

	# evaluate language choices
	language <- match.arg(language)

	variables <- make_variable_input(table_id, variables)

	call_body <- list(table = table_id,
										lang = language,
										format = "CSV",
										variables = variables)

	result <- httr::POST(DATA_ENDPOINT, body = call_body, encode = "json")

	check_http_type(result, expected_type = "text/csv")

	return(readr::read_csv2(httr::content(result, type = "text", encoding = "UTF-8"),
													locale = readr::locale(
														decimal_mark = ",",
														tz = "CET"
													)))
}


check_variables_code <- function(table_id, user_input){

	# rertieve variables for this table
	valid_vars <- get_table_metadata(table_id = table_id, variables_only = TRUE)

	# check variable names (ids)
	user_vars <- sapply(user_input, function(x) x[[1]])

	if (!all(tolower(user_vars) %in% tolower(valid_vars$id))) {
		bad_input_index <- which( !(tolower(user_vars) %in% tolower(valid_vars$id) ))
		stop(paste0("variable code '", user_vars[bad_input_index], "' is not a variable in table '", table_id,
									 "'. Run get_table_metadata('", table_id, "', variables_only = TRUE) for valid variable codes."),
						call. = FALSE)

	}

  return(TRUE)

}

check_variables_values <- function(valid_values, user_values){

	if (any(!(user_values %in% valid_values))) return(FALSE)

	return(TRUE)

}

make_variable_input <- function(table_id, user_input) {

	# Check variable codes
	check_variables_code(table_id, user_input)

	output_variable_list <- list()

	ind <- 1
	for (variable_pair in user_input) {
		# check column names of variables
		if (any(!(names(variable_pair) == c("code", "values")))) stop("variables code-values pairs need to be a named lists with names 'code' and 'values'", call. = FALSE)

		variable_id <- variable_pair$code
		variable_values <- variable_pair$values
		if (any(is.na(variable_values))) variable_values <- "*"

		valid_values <- c(get_valid_variable_values(table_id, variable_id), "*")

		if (check_variables_values(valid_values, variable_values)) {

			output_variable_list[[ind]] <- list(code = variable_id, values = I(variable_values))

		} else {

			warning(paste0("Values for ", variable_id, " are not valid... skipping ", variable_id), call. = FALSE)
			next

		}
		ind <- ind + 1
	}

	return(output_variable_list)
}
